File: v1/src/main/java/com/api/v1/services/car/CarRegistrationServiceImpl.java
package com.api.v1.services.car;

import com.api.v1.builders.car.CarBuilder;
import com.api.v1.domain.car.Car;
import com.api.v1.domain.car.CarRepository;
import com.api.v1.dtos.car.CarRegistrationRequestDto;
import com.api.v1.dtos.car.CarResponseDto;
import com.api.v1.exceptions.car.DuplicatedVinException;
import com.api.v1.mappers.car.CarResponseMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CarRegistrationServiceImpl implements CarRegistrationService {

    @Autowired
    private CarRepository repository;

    @Override
    public Mono<CarResponseDto> register(CarRegistrationRequestDto requestDto) {
        return repository
                .findAll()
                .filter(e -> e.getVin().equals(requestDto.vin()))
                .hasElements()
                .flatMap(exists -> {
                    if (exists) return Mono.error(new DuplicatedVinException(requestDto.vin()));
                    return Mono.defer(() -> {
                        Car car = CarBuilder.create().fromDto(requestDto).build();
                        car.getPlateNumbers().add(requestDto.plateNumber());
                        return repository.save(car);
                    }).flatMap(e -> Mono.just(new CarResponseMapper(e).mapToDto()));
                });
    }

}

File: v2/src/main/kotlin/com/api/v2/car/services/CarRegistrationServiceImpl.kt
package com.api.v2.car.services

import com.api.v2.car.domain.Car
import com.api.v2.car.domain.CarRepository
import com.api.v2.car.exceptions.DuplicatedVinException
import com.api.v2.car.mappers.CarResponseMapper
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class CarRegistrationServiceImpl: CarRegistrationService {

    @Autowired
    lateinit var carRepository: CarRepository

    override suspend fun register(car: @Valid Car) {
        return withContext(Dispatchers.IO) {
            if (checkTheExistenceOfDuplicatedVin(car.vin)) {
                throw DuplicatedVinException(car.vin)
            }
            val savedCar = carRepository.save(car)
            CarResponseMapper.mapToDto(savedCar)
        }
    }

    private suspend fun checkTheExistenceOfDuplicatedVin(vin: String): Boolean {
        return carRepository.findAll().filter { e -> e.vin == vin }.firstOrNull() != null
    }

}